wire/test/run-peer-wire.c

```diff
@@ -68,7 +68,7 @@ static void set_pubkey(struct pubkey *key)
 		sizeof((p1)->field)) == 0)
 
 #define eq_var(p1, p2, field)			\
-	(tal_count((p1)->field) == tal_count((p2)->field) && memcmp((p1)->field, (p2)->field, tal_count((p1)->field) * sizeof(*(p1)->field)) == 0)
+	(tal_count((p1)->field) == tal_count((p2)->field) && (((p1)->field == NULL && (p2)->field == NULL) || ((p1)->field != NULL && (p2)->field != NULL && memcmp((p1)->field, (p2)->field, tal_count((p1)->field) * sizeof(*(p1)->field)) == 0)))
```

Comment:
Maybe break this into multiple lines? It's getting a bit unwieldy.

Reply:
Just test for tal_count((p1)->field) == 0 perhaps?  tal_count(NULL) is zero.

Personally, I find the arguments against memcpy(dst, NULL, 0) to be unsatisfying, and special-casing like this has a very bad code smell.  But it'd be a really annoying bug if it ever bit us, so...

Reply:
@rustyrussell Could you expand on which arguments against `memcpy(dst, NULL, 0)` you find unsatisfying? I think the case made by Adam Langley (@agl) in [this blog post](https://www.imperialviolet.org/2016/06/26/nonnull.html) is quite compelling but I'm always open to new arguments :-)

Reply:
@jb55 @rustyrussell Would the following change be in line with your suggestions? :-)

```diff
diff --git a/wire/test/run-peer-wire.c b/wire/test/run-peer-wire.c
index 906a63f..68224c2 100644
--- a/wire/test/run-peer-wire.c
+++ b/wire/test/run-peer-wire.c
@@ -68,7 +68,8 @@ static void set_pubkey(struct pubkey *key)
                sizeof((p1)->field)) == 0)

 #define eq_var(p1, p2, field)                  \
-       (tal_count((p1)->field) == tal_count((p2)->field) && (((p1)->field == NULL && (p2)->field == NULL) || ((p1)->field != NULL && (p2)->field != NULL && memcmp((p1)->field, (p2)->field, tal_count((p1)->field) * sizeof(*(p1)->field)) == 0)))
+       (tal_count((p1)->field) == tal_count((p2)->field) && \
+               (tal_count((p1)->field) == 0 || ((p1)->field != NULL && (p2)->field != NULL && memcmp((p1)->field, (p2)->field, tal_count((p1)->field) * sizeof(*(p1)->field)) == 0)))

 /* Convenience structs for everyone! */
 struct msg_error {
```

Reply:
@rustyrussell I'm not sure how to rewrite it using a test for `tal_count((p1)->field) == 0` that would add clarity. `tal_count(NULL)` is zero but zero is also a valid size. What am I missing? :-)

Reply:
To be clear: I'm agreeing with Adam Langley.

```
#define eq_var(p1, p2, field)			\
	(tal_count((p1)->field) == tal_count((p2)->field) \
         && (tal_count((p1->field) == 0 || memcmp((p1)->field, (p2)->field, tal_len((p1)->field)) == 0))
```

(I used tal_len here, which IIRC was added to tal after this code was written).

This basically skips a 0-len memcmp, which is the problem here.